app/scripts/migrations/059.js

```diff
@@ -0,0 +1,50 @@
+import { cloneDeep, concat, groupBy, keyBy, pickBy } from 'lodash';
+import { TRANSACTION_TYPES } from '../../../shared/constants/transaction';
+
+const version = 59;
+
+/**
+ * Removes orphaned cancel and retry transactions that no longer have the
+ * original transaction in state, which results in bugs.
+ */
+export default {
+  version,
+  async migrate(originalVersionedData) {
+    const versionedData = cloneDeep(originalVersionedData);
+    versionedData.meta.version = version;
+    const state = versionedData.data;
+    versionedData.data = transformState(state);
+    return versionedData;
+  },
+};
+
+function transformState(state) {
+  const transactions = state?.TransactionController?.transactions;
+  if (
+    typeof transactions === 'object' &&
+    Array.isArray(transactions) === false &&
```

Comment:
Why the need to verify that it is not an array? And is any update to state needed when it is an array?

Reply:
Also, lodash can help simplify this:

![Screenshot from 2021-05-17 11-44-09](https://user-images.githubusercontent.com/7499938/118503514-437e4400-b705-11eb-9fb9-21f113439832.png)

Reply:
If transactions are an array here it'll mean that the state is corrupt because in a previous migration we migrated the transaction state store to an id indexed object. So if its an array this point the user has bigger problems. I added these checks to make the code as verbose as possible. There are also corresponding tests to see what would happen in the case where it isn't in the proper state.

Reply:
added isPlainObject in latest